### DISCUSSION TITLE
Kiosk mode for some users

### DISCUSSION THREAD

**Discussion by X-Dimension-88MPH** [ltsp/ltsp#820] [Q&A] [answered]
Hi! :smile: 

After switching from an outdated Ubuntu 16.04 LTSP to a new Ubuntu 22.04 LTSP setup, i want to use the kiosk mode from our old system also on the new one. 
But since we are using now GDM3 instead of LDM i don't know how to modify our old LDM guest-session script, so that i can use it with GDM3 as a postlogin oder postsession script. 

On our old LTSP server we are using this script with LDM to activate a kiosk mode for all users named "schueler*": 

case "$LDM_USERNAME" in
    schueler*)
        ssh -S "$LDM_SOCKET" "$LDM_SERVER" 'cd; rm -rf .* *; rsync -a --chmod=u=rwx,go-rwx /home/.default/ .'
        ;;
esac

I've tried to change all the LDM_ parameter in the script but with everything i do i get a black screen after the user logs in. 

Also i have read the wiki and other discussions to see how other users are setting up a kiosk/guest mode, but none of them seems to really fit to my needs because there are using nfs instead of sshfs for home directories or LightDM instead of GDM3.

So if anybody has some ideas that helps me to get it work, i would be very thankful. :blush:

**Comment by alkisg** [comment]
The correct way to set up kiosks is to override the session script.

```console
$ grep ^Exec= /usr/share/xsessions/*.desktop
Exec=mate-session
```

In my case, I'm using mate-session. So to overwrite it, I would create a /usr/local/bin/mate-session script, with contents like this:

```console
#!/bin/sh

...commands that remove stuff from home directory,
... make sure not to erase .Xauthority or other files that the DM created
...commands that copy stuff from a skeleton directory to ~
exec /usr/bin/mate-session "$@"
```

**Comment by X-Dimension-88MPH** [comment]
Thanks for your answer Alkis.

How do you start your mate-session script when the user logs on?
Is this script running with root rights at this point it starts or with the rights of the user?

Another thing i don't understand at all is how i can access the default user profile on the server from the pxe client.
As fas as i know when a user logs in its home folder is mounted via sshfs to the local client home folder. So i must use ssh like before in my new gnome-session script to copy all the things from the default profile , right?
If yes, than the next question is how do i replace the parameters "$LDM_USERNAME" "$LDM_SOCKET" "$LDM_SERVER" with the right ones?

I'm sorry but my old LDM guest-session script was not created by me so that's why i don't understand fully what it does in all of its parts.

**Comment by alkisg** [comment]
> How do you start your mate-session script when the user logs on?

It's automatically started by the display manager (e.g. lightdm). It's THE session.

> Is this script running with root rights at this point it starts or with the rights of the user?

It's running as the user.

> how i can access the default user profile on the server

One way is to store it in e.g. /etc/kiosk, inside the image.
Another way is to use `ltsp remoteapps` to mount it over sshfs.
That information already exists in discusions / issues / wiki etc, but I guess it's scattered.

Someone should create a detailed wiki page about it... unfortunately currently my free time is extremely limited.

**Comment by X-Dimension-88MPH** [comment]
Including the default profile in the image was a good idea, so i could simplify my login script to this:

case "$USER" in
    schueler*)
rsync -av --delete /etc/ltsp-kiosk/.schueler-default/ /home/"$USER"
        ;;
esac

It works fine at logout when i put these lines in /etc/gdm3/PostSession/Default for testing, but it does nothing when i create 
a file called /usr/local/bin/gnome-session like this:

#!/bin/sh
case "$USER" in
    schueler*)
rsync -av --delete --exclude '.gvfs' --exclude '.local/share/keyrings' /etc/ltsp-kiosk/.schueler-default/ /home/"$USER"
        ;;
esac

exec /usr/bin/gnome-session --session=gnome

It looks like my gnome-session script is not starting on logon. 

For testing purpose i also tired to set it in /etc/gdm3/PostLogin/Default like i do it with non-LTSP clients to enable kiosk mode, but with this setting i also get only a black screen on user logon.

**Comment by X-Dimension-88MPH** [comment, marked as answer]
Kiosk mode is now working fine for me with these steps:

1. I created a folder /etc/ltsp-kiosk 
2. In this folder i stored my default profile called .schueler-default
3. chmod -R 755 /etc/ltsp-kiosk/.schueler-default 
4. I also put my session startup skript called session-startup.sh in the /etc/ltsp-kiosk/ folder

#!/bin/sh
case "$USER" in
schueler*)
rsync -av --delete --exclude '.gvfs' --exclude '.local/share/keyrings' /etc/ltsp-kiosk/.schueler-default/ /home/"$USER"
;;
esac

exec /usr/bin/gnome-session --session=gnome

5. I have to change /usr/share/xsessions/gnome.desktop and gnome-xorg.desktop directly and change the lines

Exec=/usr/bin/gnome-session --session=gnome
to 
Exec=/etc/ltsp-kiosk/session-startup.sh

6.  ltsp image  /

Maybe it's not the best solution, but it works.
